tlog-checkpoint.md

```diff
@@ -0,0 +1,72 @@
+# Transparency Log Checkpoints
+
+A checkpoint is a [signed note][] where the body is precisely formatted for use
+in transparency log applications.  The mandatory note text includes the three
+essential parts of a log's Merkle tree head at a given size.
+
+```
+example.com/behind-the-sofa
+20852163
+CsUYapGGPo4dkMgIAUqom/Xajj7h2fB2MPA3j2jxq2I=
+
+— example.com/behind-the-sofa Az3grlgtzPICa5OS8npVmf1Myq/5IZniMp+ZJurmRDeOoRDe4URYN7u5/Zhcyv2q1gGzGku9nTo+zyWE+xeMcTOAYQ8=
+```
+
+[signed note]: https://c2sp.org/signed-note
+
+## Conventions used in this document
+
+The base64 encoding used throughout is the standard Base 64 encoding specified
+in [RFC 4648][], Section 4.
+
+`U+` followed by four hexadecimal characters denotes a Unicode codepoint, to be
+encoded in UTF-8. `0x` followed by two hexadecimal characters denotes a byte
+value in the 0-255 range. `||` denotes concatenation.
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD",
+"SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this
+document are to be interpreted as described in [BCP 14][] [RFC 2119][] [RFC
+8174][] when, and only when, they appear in all capitals, as shown here.
+
+[RFC 4648]: https://www.rfc-editor.org/rfc/rfc4648.html
+[BCP 14]: https://www.rfc-editor.org/info/bcp14
+[RFC 2119]: https://www.rfc-editor.org/rfc/rfc2119.html
+[RFC 8174]: https://www.rfc-editor.org/rfc/rfc8174.html
+[RFC 5246]: https://www.rfc-editor.org/rfc/rfc5246.html
+[RFC 6962]: https://www.rfc-editor.org/rfc/rfc6962.html
+[RFC 8032]: https://www.rfc-editor.org/rfc/rfc8032.html
+
+## Note text
+
+The note text is a sequence of at least three non-empty lines, separated by
+newlines (U+000A).
+
+ 1. The first line is the **origin**, a unique identifier for the log identity
+    which issued the checkpoint. Like note key names, it MUST be non-empty,
+    well-formed UTF-8 containing neither Unicode spaces nor plus (U+002B), and
+    it SHOULD be a schema-less URL, such as `example.com/log42`. This is only a
+    recommendation to avoid collisions, and clients MUST NOT assume that the
+    origin is following this format.
+
+ 2. The second line is the **tree size**, the ASCII decimal representation of
+    the number of leaves in the tree, with no leading zeroes (unless the tree is
+    empty, in which case the tree size is `0`).
+
+ 3. The third line is the **root hash**, the base64 encoding of the root of the
+    [RFC 6962] Merkle hash tree at the specified tree size.
+
+ 4. Any following lines are **extension lines**, opaque and optional. Extension
+    lines, if any, must be non-empty.
+
+## Signatures
+
+Logs MUST NOT sign inconsistent checkpoints, where a consistency proof can't be
+constructed from one size and root hash to the other.
```

Comment:
I think this could be tighter, as-written it would be ok for a log to sign checkpoints along multiple inconsistent views:
```
  0
 / \
A   A'
|   |
B   B'
```

`0->A` and `0->A'` are both pairwise consistent with `0` since everything is consistent with an empty tree, and `A->B` and `A'->B'` are both also pairwise consistent.

It could be more along the lines of `Logs MUST not sign any checkpoint which is inconsistent with any checkpoint it previously signed`?

Reply:
I don't see the possibility of misinterpretation in the original text that Al describes, though I do think Al's suggestion is clean and worth considering. Though would we need a definition of consistency/inconsistency? https://www.rfc-editor.org/rfc/rfc6962.html#section-2.1.2 is an option, but the language in this spec is quite different.

My bigger issue with this sentence is that `Logs MUST` introduces the concept of the Log having responsibilities. This is very reasonable, but what/who is a Log? Statements from a Log can be identified by the pair of {origin, signing key}. But as this spec mentions "key rotation" right at the end, we must assume that signing key can't be in this tuple. And origin on its own means that anyone could make checkpoints that could be inconsistent, so we must need some form of identity.

I realise that this tangent has distilled itself down to a key/identity distribution problem. But if the spec relies on there being a concept of a Log with responsibilities, do we need a definition?

In case an example helps make this real:
1. Log performs key rotation and issues checkpoint at size N with both signatures
2. At some later point it issues two checkpoints with the same origin and tree size:
  a. One signed with only the original signatures
  b. Another signed with only the new signature

If these two checkpoints have different root hashes, has the Log invalidated this rule? Or at some point did a new Log get introduced that just happened to have the same prefix as the original Log?

Reply:
Would it be reasonable to say that unless communicated via some other mechanism, a log is identified by its {origin, signing key}. And then to have an extension to this spec (or another spec) that proposes how key rotation is performed. One way that just came to mind is to use one of the extension lines for the old identity to pass the baton to the new identity. Once that is stated, then any checkpoint with the same origin signed by _either_ of these identities is considered to come from the same Log.

Reply:
Adopted Al's language, and defined consistency in terms of it being possible to construct consistency proofs, avoiding speaking separately about sizes and hashes, to Rasmus's point.

I think how to _authenticate_ a log is out of scope for this spec, and belongs in the log policy of a specific ecosystem. In the abstract sense, a log that signs inconsistent checkpoints is violating this spec. This is why I insisted on using origins to identify logs rather than keys. Whether the log violated the spec or an attacker is trying to fake a checkpoint is down to authentication.

We can definitely also write other specs for key rotation that can be optionally adopted by the ecosystems for whom it makes sense, or to interact with the witness ecosystem.